Block Deprecations: Provide extra data for isEligible check (#48815)
WordPress/gutenberg: docs/reference-guides/block-api/block-deprecation.md

-   `migrate`: (Function, Optional). A function which, given the old attributes and inner blocks is expected to return either the new attributes or a tuple array of attributes and inner blocks compatible with the block. As mentioned above, a deprecation's `migrate` will not be run if its `save` function does not return a valid block so you will need to make sure your migrations are available in all the deprecations where they are relevant.
	- _Parameters_
		- `attributes`: The block's old attributes.
		- `innerBlocks`: The block's old inner blocks.
	- _Return_
		- `Object | Array`: Either the updated block attributes or tuple array `[attributes, innerBlocks]`.
-   `isEligible`: (Function, Optional). A function which returns `true` if the deprecation can handle the block migration even if the block is valid. It is particularly useful in cases where a block is technically valid even once deprecated, but still requires updates to its attributes or inner blocks. This function is **not** called when the results of all previous deprecations' save functions were invalid.
	- _Parameters_

> 		- `attributes`: The block attributes parsed directly from the saved markup without being passed through any further code.

Reviewer: ```suggestion
		- `attributes`: The raw block attributes as parsed from the serialized HTML, and before the block type code is applied.
```

		- `innerBlocks`: The block's current inner blocks.
		- `data`: An object containing properties representing the block node and its resulting block object.

> 			- `data.blockNode`: The raw form of the block and the result of simply parsing the serialized HTML.

Reviewer: ```suggestion
			- `data.blockNode`: The raw form of the block as a result of parsing the serialized HTML.
```

> 			- `data.block`: The block object, which is the result of feeding the `blockNode` through the block type's code.

Reviewer: ```suggestion
			- `data.block`: The block object, which is the result of applying the block type to the `blockNode`.
```

-   `supports` (Object): The [supports definition](/docs/reference-guides/block-api/block-registration.md) of the deprecated form of the block.
-   `save` (Function): The [save implementation](/docs/reference-guides/block-api/block-edit-save.md) of the deprecated form of the block.
-   `migrate` (Function, Optional): A function which, given the old attributes and inner blocks is expected to return either the new attributes or a tuple array of `[ attributes, innerBlocks ]` compatible with the block. As mentioned above, a deprecation's `migrate` will not be run if its `save` function does not return a valid block so you will need to make sure your migrations are available in all the deprecations where they are relevant.

> -   `isEligible` (Function, Optional): A function which, given the attributes and inner blocks of the parsed block, returns true if the deprecation can handle the block migration even if the block is valid. This is particularly useful in cases where a block is technically valid even once deprecated, but still requires updates to its attributes or inner blocks. If further information is required to determine this deprecation's eligibility, the block node and object are provided within a third parameter. This function is not called when the results of all previous deprecations' `save` functions were invalid.

Reviewer: What do you think about re-structuring the text for `isEligible` (and maybe `migrate` too)?

A few changes that I think would be good:
- Document the arguments using bullet points (matching what we do for function that have JSDoc types).
- Detail the expected return type (again like we do for JSDoc types).
- Ensure the terms we're using are consistent (e.g. parsed block and block node are both mentioned currently).
- Add some more clarity to that last sentence. I've read it a few times and still don't understand it.

Ideally `isEligible` and `migrate` are so complicated as concepts that they should be in a separate section, but that's not one for this PR.

Author: Some great suggestions there, thank you 👍 

I've tried reformatting and rewording the `isEligible` and `migrate` function docs in https://github.com/WordPress/gutenberg/pull/48815/commits/efd2cd7bd6f484ada942433ed4fc32dc4786e261. 

Let me know if any further tweaks are needed in this PR.
